lib/pbench/server/auth/__init__.py

```diff
@@ -0,0 +1,435 @@
+            {
+                "aud": "client_id",
+                "email_verified": "true",
+                "expires_in": "3572",
+                "access_type": "offline",
+                "exp": "1660087224",
+                "azp": "client_id",
+                "scope": "openid email profile",
+                "email": "email",
+                "sub": "106203657987367771589"
+            }
+        """
+        payload = {
+            "client_id": self.client_id,
+            "client_secret": self.client_secret_key,
+            "token": token,
+        }
+        return self._post(token_info_uri, data=payload).json()
+
+    def token_introspect_offline(
+        self,
+        token: str,
+        key: str,
+        audience: str = "account",
+        algorithms: List[str] = ["RS256"],
+        **kwargs,
+    ) -> JSON:
+        """
+        Utility method to decode access/Id tokens using the public key provided by the identity provider
+        The introspected JWT token contains the claims specified in https://tools.ietf.org/html/rfc7662
+        Args:
+            token: token value to introspect
+            key: client public key
+            audience: jwt token audience/client, who this token was intended for
+            algorithms: Algorithm with which this JWT token was encoded
+        Returns:
+            Extracted token information
+            {
+                "aud": "client_id",
+                "email_verified": "true",
+                "expires_in": "3572",
+                "access_type": "offline",
+                "exp": "1660087224",
+                "azp": "client_id",
+                "scope": "openid email profile",
+                "email": "email",
+                "sub": "106203657987367771589"
+            }
+        """
+        return jwt.decode(
+            token, key, algorithms=algorithms, audience=audience, **kwargs
+        )
+
+    def get_userinfo(self, token: str = None) -> JSON:
+        """
+        The userinfo endpoint returns standard claims about the authenticated user,
+        and is protected by a bearer token.
+        http://openid.net/specs/openid-connect-core-1_0.html#UserInfo
+        Args:
+            token: Valid token to extract the userinfo
+        Returns:
+            Userinfo payload
+            {
+                "family_name": "family_name",
+                "sub": "106203657987367771589",
+                "picture": "profile_picture_link",
+                "locale": "en",
+                "email_verified": true,
+                "given_name": "given_name",
+                "email": "email",
+                "hd": "redhat.com",
+                "name": "full_name"
+            }
+        """
+
+        if token:
+            self.add_header_param("Authorization", f"Bearer {token}")
+
+        return self._get(self.USERINFO_ENDPOINT).json()
+
+    def logout(self, refresh_token: str) -> HTTPStatus:
+        """
+        The logout endpoint logs out the authenticated user.
```

Comment:
We still need to decide how we want to write those top-level callers of these functionalities and how much access we need to give to the end user to deal with OIDC third-party clients from Pbench-server. You correctly pointed out somewhere else that we don't have a good use case for all of these functionalities yet, but these functions provide basic token handling capabilities for the user if they are logged in with our Keycloak server.  

The logout and revoke functionalities might be useful for CLI users. For example, one of the features that I am working on is `token_exchange`, consider the following scenario:
1. User logs in (either with Pbench or some other identity provider)
2. They get the initial access and refresh token
3. User wants to access certain restricted Pbench endpoints which need Pbench tokens (separate from the above tokens)
4. User uses the identity access token to exchange it with the "Pbench" access and refresh token
5. "Pbench" tokens specific to Pbench, can only be revoked or refreshed by individuals by logging into the Keycloak server, or in this case, we can provide REST endpoints to do the same if we don't want to give users access to our Keycloak server.  

Also, granted that the third-party access tokens can be revoked by the user by logging into the identity providers token management site but I do think Pbench as a client should be able to invalidate these tokens (as they are provided for the Pbench client) in case the user wants to. 

The logout REST endpoint is not provided by most OIDC clients like google/GitHub but I have not confirmed whether there are any that do provide it. So if it is not provided we can not log out of the user session from the Pbench Server anyway, if someone tries to do it we will abort the operation with `METHOD_NOT_ALLOWED`.

Reply:
I would like to see if we can simplify this situation.

I propose that the Pbench Server _not_ provide services for creating and maintaining tokens.  This should be done by a separate, "external" service, although, if we need to provide for "local" logins, we'll also provide an instance of the service, but it can be separate from the Pbench Server.  The user, via either the GUI or CLI, negotiates with the external IAM service for a token and presents that to the Pbench Server; the Server then validates the token, and proceeds accordingly.  If/when the token expires, the user (i.e., the GUI or CLI) negotiates a renewal with the IAM service.  I don't know how logout works, but with short-duration tokens, it's probably not important.  (And, for pushing results to the Pbench Server, I'm assuming that we will use an "access key", which is a separate beast.)  So, with this as a foundation, why would the Pbench Server need to create tokens itself?

I can imagine situations in which it would be convenient for us to have our "own" tokens, but I'm thinking that it's not worth the hassle.  (The problem of naming the d*mn things alone is formidable.)  So, if we can get by without them, I think it's worth seriously considering.

So, to work through your flow above,

1. The user authenticates with an external IAM service (one which the Pbench Server supports, and this list will presumably be generated by an API call to the Pbench Server) 
2. The user receives an access token and a refresh token from the external IAM service
3. The user wants to access a Pbench service which requires authentication
4. The user presents the externally-obtained access token to the Pbench Server which verifies it
5. There is no need for the user to interact with the Pbench Server to renew or revoke the access token.

If the Pbench Server can validate an external token and use it to authenticate a user, why do we need to add our own tokens to the mix?

And, given that the user already negotiates the initial token directly with the external IAM provider, why would we want to interpose the Pbench Server between them for things like token invalidation?  (Presumably we want to provide support like that in the _Dashboard_, but why would we need a Server API for that?)

Reply:
There is nothing wrong with the approach that you suggested where the Pbench server receives an access token from the user and if the server can validate it, it will grant access to the user. We would have certain endpoints that would need only this level of authentication. The issue arises when we try to restrict access based on the user roles/groups attributes. We don't get any role association from the third-party tokens. 

To do that we would either have to 
1. Get the user id by decoding the third-party access token, query that id with our Keycloak server where all the user entries are stored to retrieve the role/group association. It is likely that we might have to hit our keycloak server or our user database multiple times to fetch that information based on how many times the user tries to access our endpoints (unless we implement some sort of cache but that's a different point). 
2. We just give them the Pbench-specific access token (very short-lived) and a refresh token (a very long life token) that has other meta information encoded related to the user, which if presented we should be able to decode it without hitting the server or database. This is handy in generating tokens with the specific role assigned for example if the token has a role `result_push` then only that token can be used for pushing results on the server and something like that. And with a very long life span refresh token, a user can refresh the access token as many times as he/she wants without generating another third-party identity token.

Reply:
I'm pretty sure that we're going to want to address the "results push" with a completely different type of token (i.e., an "access key" which does not expire), so we should leave that aside.

As for determining roles, I don't think that we're going to want to encode that into the token.  We're going to need to be cognizant of how much we put into the token -- we don't want it to get too big -- and I'm not sure that we're going to want to bound the number of roles and groups that a user can have or be a member of.  So, I suspect that membership and role application is going to be a database transaction.  Thus, all we need from the token is the user's identity.

Reply:
But if we already have role management then why do we need a separate mechanism for generating a very specific token in this case for `results push` that can be created by the same mechanism by just embedding a specific role in it? I am saying if we include the role in the token then we don't need a separate mechanism for creating a `results push` token.

Tokens getting too large is a valid concern, but I am sure we can limit which client roles should be included in the token and which client roles should not. If certain server endpoints require some rare role assignment, it can be queried with database transactions if not found in the token locally.

Reply:
Push tokens are likely to be permanent (or _very_ long lived...existing until revoked).  They'll need to be unguessable, but, so long as the server has a way to recognize them, they don't even _have_ to contain identity information.  If they are generated by the server, they can be almost arbitrary in value -- the server just stores them and matches them when they are supplied by the request.  (If the request includes an identity which we can use as a key, we can look up the stored copy of the push token, but there are other options for finding it.)

If we go with my assertion that we should only use an external IAM service, then the Pbench Server won't have another mechanism for generating tokens (internally), so there won't be a "same mechanism" that we would want to somehow extend for push tokens.

There are lots of examples of this sort of thing (AWS "API keys", COPR "API tokens", Quay.io "robot account tokens", et al.)  An authenticated user would be able to request the creation of a push token, or request a copy of an existing one, which they could store on the system from which they want to do pushes.  And, presumably, we would allow users to delete or regenerate their push tokens.

The details of what is in a push token or how one is generated by the server isn't that important.  What is important is the token is unguessable and that the server can validate it.  It will be obvious from context where it is used that it is a push token, so it won't be confused with other types of tokens.  So, I think we can treat push tokens separately from access tokens used in other parts of the API.  And, in any case, I don't think we need to worry about them in the context of this PR.

Reply:
Yeah, I think it is better to discuss this topic in a group discussion. But I thought we agreed to have separate tokens for accessing Pbench server endpoint that is different from the IDP tokens even for generally accessing all the endpoints not just the results push isn't it?

Reply:
Another group discussion would likely be good.  (I think the last one ended with an agreement to update the flow diagram before proceeding to the down-stream details.  🤣)